### ISSUE TITLE
Make badges accessible / friendly for blind, visually impaired users and screenreaders

### ISSUE THREAD

**Issue by leonstafford** [emibcn/badge-action#2]
Hi @emibcn, this is just what I was looking for to control my own badges. The main driver, though, is to provide better #a11y than all the badges out there, which lack proper markup for screen readers, which visually impaired users rely on.

Are you interested to build in/accept a PR to add that to this project? Possibly, with an option to just use textual badges (gradient badges are pretty, but also eats up extra bandwidth and the real important information doesn't need an image).

**Comment by emibcn** [comment]
Of course I am open to make it better. But be aware that this action is just a wrapper around a JavaScript library ([gradient-badge](https://github.com/bokub/gradient-badge/)) which modifies a SVG created by a dependency of that one ([badgen](https://github.com/badgen/badgen)).

The created SVG contains the text in it also inside of an `aria-label` attribute in the main `<svg>` element, which will be read by a screen reader, as far as I know.

If the question is about removing all the SVG code dedicated to colors and gradients, we can see if we can ask for this to the people who maintains the other two libraries. If they refuse to do so, we can try to modify the SVG removing the unwanted parts, of course 😉

**Comment by leonstafford** [comment]
Awesome, thanks for response @emibcn!

I looked into it more and started my own SVG-less badge action, but will link back to this and see about upstream lib improvements.

**Comment by leonstafford** [comment]
https://github.com/leonstafford/a11y-friendly-badges

**Comment by leonstafford** [comment]
@emibcn bit more on the SVG situation here: https://twitter.com/fransallen/status/1357197241199431683

So my thoughts are that the SVG extra markup is more of a band-aid and would really need the useful information parsable from within it, ie the code coverage %/important metric.

I just don't see the need for any images when text works fine and is accessibility-first, not an afterthought.

I'm designing my action to be flexible with any scripts passed in, which should lead to having a bunch of example scripts for common tooling one would want to extract stats from.

**Comment by emibcn** [comment]
The interesting thing about SVG images is that it can be updated without updating the document containing it (ie, `README.md`). For example, you can push new tests which trigger changes on test coverage and it will automatically update the badge, without needing a new `README.md` version.

Does that makes sense to you?

Some solutions for improving accessibility without adding extra commits to `README.md`: 
- Maybe a good approach would be to modify the screen readers so them can extract some metadata from the images (in the case of SVG, its XML, like HTML, so it wouldn't be that hard).
- Another approach would be to ask GitHub (and many other providers) to allow some text including/templating, where we could add textual information.
-  Another bigger approach, would be to ask the [W3C](https://www.w3.org/) to add a new tag allowing to import textual information. Maybe something similar to `<img src="something.svg" />`, but for external plain text instead: `<textual src="something.txt" />`.

**Comment by leonstafford** [comment]
Yeah, I think that's the biggest hurdle with this - it's no longer just copy paste some markdown for the developers to get the stats updated. It should still be easy, especially with templated scripts. But this is where a11y efforts need to be improved and developers are problem-solvers, we just need to care more about being properly inclusive, especially if "open source".

Re the ideas:


>    Maybe a good approach would be to modify the screen readers so them can extract some metadata from the images (in the case of SVG, its XML, like HTML, so it wouldn't be that hard).

I'm of opinion that screenreaders shouldn't need to advance. We used to build websites that worked fine in screenreaders, using text and anchors, we just started to break things horribly, to detriment of accessibility and performance. If it doesn't render in the lynx/links browsers cleanly, it's no good.

>    Another approach would be to ask GitHub (and many other providers) to allow some text including/templating, where we could add textual information.

Would be great, but then we're vendor-locked into GitHub and the README's should still be standalone usable within a repo offline or hosted in own git (ie, good luck using GitHub if you're from Iran or other nation on US's sh*t-list). For same reason, this kind of action shouldn't just be built and stop when it works in GitHub, but attacking that first, as it's the most popular. Could be ported to work in CircleCI, Travis, etc.

> Another bigger approach, would be to ask the W3C to add a new tag allowing to import textual information. Maybe something similar to <img src="something.svg" />, but for external plain text instead: <textual src="something.txt" />.

Similar to first point, I'd go for accessibility first approach, not extra markup when simple HTML already works fine.

I'm new and inexperienced in a11y stuff and have an idealistic view that we can get a lot of the way there just by keeping things clean/minimalist. No need to compromise on the content, but let's get rid of the "pretty" stuff that breaks things for many users.

If @rkingett has time for input, he may tell me if I'm being too idealistic/extreme and if waiting for screen-readers to get better/adding extra markup to things makes more sense.

**Comment by rkingett** [comment]
I am strongly against waiting on screen readers to advance because, honestly, that should not be their job. Plus, making it more minimalist and inclusive just makes things better for people running different browsers, not just screen readers. I would rather screen reader manufacturers worry about extra navigational aids rather than making something accessible. The developer is in a position to make things better for everyone. Not everybody uses screen readers, though.

**Comment by emibcn** [comment]
This action is for creating badges. I can help making them more accessible, but I can't help making them disappear, at least not from here. If you have any suggestion for making the badges more accessible, please, concretize it and I will do my best, even mocking the dependent libraries if its needed.

Even more, If you ask for it, I can create a new action for modifying `README.md` directly with a conditional action (ie, prevent infinite actions loop). But, in my opinion, very little devs will agree an action committing and pushing against their `main`/`master` branch for every single push (double the pushes, probably also doubling the commits).

**Comment by leonstafford** [comment]
Yep, it's a tall order. I'll continue with my text badges and ponder easier options. Thanks for all your discussion on this, it's a great challenge to work on!

**Comment by emibcn** [comment]
OK. Then I close the issue. If you need anything, please, don't hesitate to ask for it.
